### DISCUSSION TITLE
"Don't Ask to Ask" and "XY Problem"

### DISCUSSION THREAD

**Discussion by ghost** [QuiltMC/developer-wiki#110] [General]
In regards to: https://wiki.quiltmc.org/en/introduction/getting-started

> Asking questions properly, whether it is on our forum or on Discord, is an essential skill to have. Never be afraid to ask questions, we’re always willing to help, and we don’t judge.
> 
> When asking questions, try to include the necessary context in your first message, and [do not ask to ask](https://dontasktoask.com/). State your problem, what you have already tried to solve it, and how those attempts failed. Make sure to avoid the [XY problem](https://xyproblem.info/) by not only asking for help for one specific part of a solution to your problem, but also stating the actual problem you are trying to solve.

Just wanted to point out a few problems with the language of this page on the wiki: it's _condescending_. Dictating the specific particular ways people have to ask questions is pretentious when all that really matters is that a question was asked and someone is able to answer it. Even the link to the page explaining why you apparently "shouldn't ask to ask" is presumptuous and pretentious. We are not _divas_, we are programmers. This "bad form" is made up. In fact, I've never seen anyone say something like this is "bad form" in the many many programming communities that I have been in. Really, it's just extremely petty.

As for the XY problem, that's just as pretentious because it's the excuse StackOverflow users use to berate people who ask a question about a particular problem and someone can't imagine how it might be useful, so they automatically determine that it's not useful to begin with, as if that person knows every single thing about what solutions are useful for particular problems. Again, it's pretentious and condescending, and even the examples given in this XY problem website are _intentionally convoluted and unrealistic_. The biggest problem with this "XY Problem" is that some people think that Y is not a valid solution, whereas many others would say it _is_ a valid solution. Ultimately, you answer what you've been asked because that's what you've been asked to help with. It is up to the asker to determine whether the solution is valid or not, or to ask whether it's a valid solution or not. Nothing more, nothing less. StackOverflow is literally _notorious_ for these types of responses, and for just generally terrible solutions in general. StackOverflow should never be used as a model for this type of stuff, and it is not good to advocate that type of behavior in a programming community. I can speak more on why this XY Problem thing is actually *extremely* bad for interactions *and learning* in general if people wish, but I think I got the main point across.

**Comment by ghost** [comment]
As for fixing the language so it's not so condescending, here's my solution:

> Asking questions properly, whether it is on our forum or on Discord, is an essential skill to have.

This is obvious, people already know this. You don't have to make them feel dumb by suggesting this is some new-fangled idea that you came up with that is *really important* to get across. Remove this line completely, it's not necessary. It's probably *the* most condescending line on this entire page.

> Never be afraid to ask questions, we’re always willing to help, and we don’t judge.

This is a great line! Keep it! But it contradicts the previous line (you say you won't judge, but you are already pre-judging them by telling them about the precise way they have to ask questions, lmao), which will no longer be a problem when that previous line is removed.

> When asking questions, try to include the necessary context in your first message, and [do not ask to ask](https://dontasktoask.com/).

Uh, no. There's nothing wrong with asking to ask, and it doesn't imply anything except what users mistakenly read into it. Remove the do not ask to ask, it makes y'all sound pretentious. You can keep the first part though, because it's generally good advice.

> State your problem, what you have already tried to solve it, and how those attempts failed. Make sure to avoid the [XY problem](https://xyproblem.info/) by not only asking for help for one specific part of a solution to your problem, but also stating the actual problem you are trying to solve.

The XY problem stuff can be removed. It's unnecessary and even the website you link to is actually just really rude. You can say almost the same thing without sounding so pretentious by just saying something more like this: "Be sure to state your problem, what you have already tried to solve it, and how those attempts failed. Having the full context of the problem helps others guide you to the right solution".

This is how the entire thing should read, according to my suggestions above:
> Never be afraid to ask questions, we’re always willing to help, and we don’t judge. When asking questions, try to include the necessary context in your first message. Be sure to state your problem, what you have already tried to solve it, and how those attempts failed. Having the full context of the problem helps others guide you to the right solution.

There, simple, easy to understand, not condescending or presumptuous. Much better, imho.

**Reply by ghost** [reply]
A few more suggestions for this page:

> When debugging, include your latest.log and when your game crashes, a crash log. Additionally, full access to your source code can be really helpful, ideally uploaded to a website such as GitHub.

These lines are perfectly fine.

> There are many topics in modding where you quickly will find yourself at a point where there are no tutorials to help.

Grammar edit: There are many topics in modding where you _will_ quickly find there are no tutorials to help.
Further improvement: There are many topics in modding where you will quickly find there are no _relevant_ tutorials.
Another potential improvement: You may quickly find there are many topics in modding lacking relevant tutorials.

> In those cases you’ll usually either have to understand the relevant Minecraft code, or look at other mods and see how they got similar things done. 

You don't have to understand the Minecraft code until you read the code, really, so I disagree with the first part of this sentence quite a lot. Here's how I would rewrite this: "In those cases, you may have to read the relevant Minecraft code, or look at other mods and see how they implemented similar ideas." (or similar functions, or similar solutions, etc.)
It's not the best to point out in the text of your writing that readers might be unfamiliar with certain things; things like "it's easy", "you should understand", "you probably don't know this but...", or pointing out obvious knowledge that any person would reasonably know (like the value of asking questions well) are generally things you don't want to include in writing like this. Don't put them in tutorials, don't put them into wikis, don't put them in books, etc.

The biggest thing in writing is to know your audience. Those who are getting into modding, or are beginner modders, might very well already know things about programming in general. Some of them might have even programmed for a decade or more. Know your audience.

**Comment by gdude2002** [comment]
(If you're wondering why this is being commented on now, someone decided to bring it up!)

While I think that the pages supposedly explaining "don't ask to ask" and the XY problem are a poor way to direct users on how to ask questions, I think it's important to note that neither concept is particularly problematic on its own.

Many communities state "don't ask to ask" largely because that's a bad way to get help - anyone providing help in a community space is a volunteer with limited time, and getting to the point by asking your question immediately is far more likely to encourage others to actually answer your question, in my experience. That said, I don't think the link used is particularly helpful - it'd be better to use some nicer phrasing in the wiki itself, IMO.

As for the XY problem, it absolutely is a real problem that I run into all the time in programming communities. I agree that the link sucks and things should be rephrased, but disagree that the concept only exists for Stack Overflow users to be condescending (god knows they don't need an excuse to do that). I do also think it's useful to include that specific term on the page, since it's something many would-be helpers are likely to be familiar with and potentially bring up when explaining how they can better ask their questions next time.

I think it's also worth noting that _knowing how to ask for help is a skill that many people haven't yet honed._  Instead of removing the implication that it's important to hone this skill, I think the wiki would be better served by creating a page with advice on how to better ask for help, the information that helpers will require and expect (and why that is), written in a positive and encouraging tone. This is something that worked well when I was a co-owner at Python Discord, and [they still have an expanded version of the page I wrote today](https://www.pythondiscord.com/pages/guides/pydis-guides/asking-good-questions/) - though it doesn't have the same optimistic tone that I previously used.

I'd also just like to point out that those helping out are people too, and there is an implied level of etiquette that many users seeking support don't try to reach. While I don't think it's useful to talk about "help vampires" and other such terms, I think the wiki should at least ask for some patience and mutual respect.

---

Out of curiosity, why isn't this a PR?
